fenzo-core/src/main/java/com/netflix/fenzo/InternalVMCloner.java

```diff
@@ -0,0 +1,223 @@
+                    for (String k: maxScalars.keySet())
+                        scalars.compute(k, (s, oldVal) -> oldVal + maxScalars.get(k));
+                }
+                final Map<String, Protos.Attribute> attrs = avm.getCurrTotalLease().getAttributeMap();
+                if (attrs != null && !attrs.isEmpty()) {
+                    for (Map.Entry<String, Protos.Attribute> e: attrs.entrySet())
+                        attributeMap.putIfAbsent(e.getKey(), e.getValue());
+                }
+            }
+        }
+        final double fcpus = cpus;
+        final double fmem = mem;
+        final double fdisk = disk;
+        final double fnetwork = network;
+        final List<VirtualMachineLease.Range> fports = Collections.singletonList(
+                new VirtualMachineLease.Range(100, 100 + (int)ports));
+        return new VirtualMachineLease() {
+            @Override
+            public String getId() {
+                return "NoID";
+            }
+
+            @Override
+            public long getOfferedTime() {
+                return 0;
+            }
+
+            @Override
+            public String hostname() {
+                return "NoHostname";
+            }
+
+            @Override
+            public String getVMID() {
+                return "NoID";
+            }
+
+            @Override
+            public double cpuCores() {
+                return fcpus;
+            }
+
+            @Override
+            public double memoryMB() {
+                return fmem;
+            }
+
+            @Override
+            public double networkMbps() {
+                return fnetwork;
+            }
+
+            @Override
+            public double diskMB() {
+                return fdisk;
+            }
+
+            @Override
+            public List<Range> portRanges() {
+                return fports;
+            }
+
+            @Override
+            public Protos.Offer getOffer() {
+                return null;
+            }
+
+            @Override
+            public Map<String, Protos.Attribute> getAttributeMap() {
+                return attributeMap;
+            }
+
+            @Override
+            public Double getScalarValue(String name) {
+                return scalars.get(name);
+            }
+
+            @Override
+            public Map<String, Double> getScalarValues() {
+                return scalars;
+            }
+        };
+    }
+
+    /* package */ VirtualMachineLease cloneLease(VirtualMachineLease lease, String hostname, long now) {
+        final List<VirtualMachineLease.Range> ports = new LinkedList<>();
+        final List<VirtualMachineLease.Range> ranges = lease.portRanges();
+        if (ranges != null && !ranges.isEmpty()) {
+            for (VirtualMachineLease.Range r: ranges)
+                ports.add(new VirtualMachineLease.Range(r.getBeg(), r.getEnd()));
+        }
+        return new VirtualMachineLease() {
```

Comment:
Some fields in 'lease' object are mutable. It might be safer to make a full copy, unless thee is a guarantee these values will not change during shortfall analysis.